File: sensei-bryan/netlify/functions/messages.py
from openai import OpenAI
import json
import os

#meta-llama/llama-3.2-1b-instruct:free
#              ^^ Free but terrible LLM model - Definitely use for testing however

# openai/gpt-4.1-nano
#          ^^ Not free, but affordable and much better preformance

MODEL = "mistralai/mistral-nemo"


# def get_response(prompt):
   
#     print("......")
    
#     completion = client.chat.completions.create(
#         model = MODEL,
#         messages=[{"role": "user","content": prompt}]
#     )

#     print(completion.choices[0].message.content)

#     print("\n\n") # temporary spacing for demo
#     return False

def handler(event, context):
    # Handle CORS
    headers = {
        'Access-Control-Allow-Origin': '*',
        'Access-Control-Allow-Headers': 'Content-Type',
        'Access-Control-Allow-Methods': 'POST, OPTIONS'
    }
    
    # Handle preflight requests
    if event['httpMethod'] == 'OPTIONS':
        return {
            'statusCode': 200,
            'headers': headers,
            'body': ''
        }
    
    if event['httpMethod'] != 'POST':
        return {
            'statusCode': 405,
            'headers': headers,
            'body': json.dumps({'error': 'Method not allowed'})
        }
    
    try:
        # Parse the request body
        body = json.loads(event['body'])
        input_text = body.get('text', '')
        
        if not input_text:
            return {
                'statusCode': 400,
                'headers': headers,
                'body': json.dumps({'error': 'No text provided'})
            }

        client = OpenAI(base_url="https://openrouter.ai/api/v1", api_key=os.environ.get('ai_auth'))
        response = client.chat.completions.create(
            model=MODEL,
            messages=[{"role": "user", "content": input_text}]
        )
        
        result = response.choices[0].message.content
        
        return {
            'statusCode': 200,
            'headers': headers,
            'body': json.dumps({'result': result})
        }
        
    except Exception as e:
        return {
            'statusCode': 500,
            'headers': headers,
            'body': json.dumps({'error': str(e)})
        }

File: sensei-bryan/netlify/functions/messages.ts

/**
 * Netlify Function: Chat Message Handler
 * 
 * This serverless function handles chat interactions between users and the AI model.
 * It processes incoming messages, sends them to OpenRouter's API, and returns the AI's response.
 * 
 * Environment Variables Required:
 * - auth_ai: OpenRouter API key
 * - OPENROUTER_API_URL: OpenRouter API endpoint
 * - model: AI model to use (e.g., "mistralai/mistral-nemo")
 */

import type { Handler } from '@netlify/functions';


/**
 * Main handler function for processing chat messages
 * @param {Object} event - The incoming HTTP event from Netlify
 * @returns {Promise<Object>} Response object with status code and message
 */
export const handler: Handler = async (event) => {

	//hanfle wrong kind of request ( not POST)
	if (event.httpMethod !== 'POST') {
		return {
			statusCode: 405,
			body: JSON.stringify({ error: 'Method Not Allowed' }),
		};
	}


	try {
		const { text } = JSON.parse(event.body || '{}');
		if (!text) {
			return {
				statusCode: 400,
				body: JSON.stringify({ error: 'No message provided' }),
			};
		}

		const apiKey = process.env['auth_ai'];
		const apiURL = process.env['OPENROUTER_API_URL'];
		const model = process.env['model'];

		if (!apiKey) {
			return {
				statusCode: 500,
				body: JSON.stringify({ error: 'Missing OpenRouter API key' }),
			};
		}

		const response = await fetch(apiURL || '', {
			method: 'POST',
			headers: {
				'Content-Type': 'application/json',
				'Authorization': `Bearer ${apiKey}`,
			},
			body: JSON.stringify({
				model: model,
				messages: [
					{ role: 'system', content: 'You are Sensei Bryan, a helpful educational AI assistant.' },
					{ role: 'user', content: text },
				],
			}),
		});

		if (!response.ok) {
			const errorText = await response.text();
			return {
				statusCode: response.status,
				body: JSON.stringify({ error: errorText }),
			};
		}

		const data = await response.json();
		const result = data.choices?.[0]?.message?.content || 'No response from AI.';

		return {
			statusCode: 200,
			body: JSON.stringify({ result }),
		};
	} catch (error: any) {
		return {
			statusCode: 500,
			body: JSON.stringify({ error: error.message || 'Unknown error' }),
		};
	}
};
